File: python/unique_paths.py
class Solution:
#     # brute force dfs
#     def uniquePaths(self, m: int, n: int) -> int:
#         # subproblem: how many ways to reach finish from any point on grid?

#         # out of bounds, so invalid path
#         if m < 1 or n < 1:
#             return 0

#         # finish reached from valid path 
#         if m == 1 and n == 1:
#             return 1

#         return self.uniquePaths(m - 1, n) + self.uniquePaths(m, n - 1)

    # optimized memoized dfs
    def uniquePaths(self, m: int, n: int) -> int:
        return self.countUniquePaths(m, n, {})

    def countUniquePaths(self, row: int, col: int, cache: dict[tuple[int, int], int]) -> int:
        # subproblem: how many ways to reach finish from any point on grid?

        # hit cache to get path count
        if (row, col) in cache:
            return cache[(row, col)]

        # out of bounds, so invalid path
        if row < 1 or col < 1:
            return 0

        # finish reached from valid path 
        if row == 1 and col == 1:
            return 1

        pathCountFromCell = self.countUniquePaths(row - 1, col, cache) + self.countUniquePaths(row, col - 1, cache)

        # cache path count to finish from current cell
        cache[(row, col)] = pathCountFromCell
        return pathCountFromCell

File: cpp/unique_paths.cpp
class Solution {
public:
    // optimal top down dp approach
    int uniquePaths(int m, int n) {
        if (m == 0 || n == 0) {
            return 0;
        }
        vector<vector<int>> cache(m, vector<int>(n, -1));
        return countPathsFrom(cache, m - 1, n - 1);
    }

    int countPathsFrom(vector<vector<int>> &cache, int row, int col) {
        // reached out of bound cell by going up or left too far
        if (row < 0 || col < 0) {
            return 0;
        }
        // finish cell reached from a valid path
        if (row == 0 && col == 0) {
            return 1;
        }
        if (cache[row][col] != -1) {
            return cache[row][col];
        }

        // number of ways to reach finish cell from current cell is the sum
        // of number of ways to reach finish from left and above cells
        int curTotalPaths = countPathsFrom(cache, row, col - 1) +
                            countPathsFrom(cache, row - 1, col);

        cache[row][col] = curTotalPaths;
        return curTotalPaths;
    }
};
